compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -521,49 +521,92 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
+                    | ty::Char
+                    | ty::Int(_)
+                    | ty::Uint(_)
+                    | ty::Float(_)
+                    | ty::Adt(_, _)
+                    | ty::Foreign(_)
+                    | ty::Str
+                    | ty::Array(_, _)
+                    | ty::Slice(_)
+                    | ty::RawPtr(_)
+                    | ty::Ref(_, _, _)
+                    | ty::FnDef(_, _)
+                    | ty::FnPtr(_)
+                    | ty::Dynamic(..)
+                    | ty::Closure(..)
+                    | ty::Coroutine(..)
+                    | ty::CoroutineWitness(..)
+                    | ty::Never
+                    | ty::Tuple(_)
+                    | ty::Param(_)
+                    | ty::Placeholder(..)
+                    | ty::Infer(ty::IntVar(_) | ty::FloatVar(_))
+                    | ty::Alias(ty::Inherent, _)
+                    | ty::Alias(ty::Weak, _)
+                    | ty::Error(_) => break,
+                    ty::Infer(ty::FreshTy(_) | ty::FreshIntTy(_) | ty::FreshFloatTy(_))
+                    | ty::Bound(..) => bug!("unexpected self type for `{goal:?}`"),
+
+                    // If we hit infer when normalizing the self type of an alias,
+                    // then bail with ambiguity.
+                    ty::Infer(ty::TyVar(_)) => {
+                        if let Ok(result) = ecx
+                            .evaluate_added_goals_and_make_canonical_response(Certainty::AMBIGUOUS)
+                        {
+                            candidates
+                                .push(Candidate { source: CandidateSource::AliasBound, result });
+                        }
+                        break;
+                    }
 
-        for assumption in
-            self.tcx().item_bounds(alias_ty.def_id).instantiate(self.tcx(), alias_ty.args)
-        {
-            match G::consider_alias_bound_candidate(self, goal, assumption) {
-                Ok(result) => {
-                    candidates.push(Candidate { source: CandidateSource::AliasBound, result })
+                    // Excluding IATs and type aliases here as they don't have meaningful item bounds.
+                    ty::Alias(kind @ (ty::Projection | ty::Opaque), alias_ty) => (kind, alias_ty),
+                };
+
+                for assumption in
+                    ecx.tcx().item_bounds(alias_ty.def_id).instantiate(ecx.tcx(), alias_ty.args)
+                {
+                    match G::consider_alias_bound_candidate(ecx, goal, assumption) {
+                        Ok(result) => {
+                            candidates
+                                .push(Candidate { source: CandidateSource::AliasBound, result });
+                        }
+                        Err(NoSolution) => {}
+                    }
+                }
+
+                // If we have a projection, check that its self type is a rigid projection.
+                // If so, continue searching.
```

Comment:
please elaborate here here and point to an example test

Reply:
or :thinking: didn't see the comment at the start of the loop

style: can you change this to a recursive function taking the current self type and move all the documentation to the method itself?

Reply:
TODO: still need to do this